lib/SILGen/SILGenLValue.cpp

```diff
@@ -3930,6 +3946,16 @@ LValue SILGenLValue::visitMemberRefExpr(MemberRefExpr *e,
     }
   }
   
+  // if we are trying to access an init accessor of a property out of the
+  // lexical type context, emit an error.
+  if (strategy.hasAccessor() && strategy.getAccessor() == AccessorKind::Init) {
+    AccessorDecl* initAccessorDecl = var->getOpaqueAccessor(AccessorKind::Init);
+    if (initAccessorDecl && !isInitAccessorAccessible(SGF, initAccessorDecl)) {
```

Comment:
I think this needs to happen during Sema, `Init` based strategy is only produced for immutable things and it should be possible to prevent that from happening, we have a few places were we pretend that property is mutable so the availability check should happen there.